src/coreclr/jit/lclmorph.cpp

```diff
@@ -1166,41 +1166,76 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
 #ifdef FEATURE_HW_INTRINSICS
             if (varTypeIsSIMD(varDsc))
             {
-                // We have three cases we want to handle:
-                // 1. Vector2/3/4 and Quaternion where we have 4x float fields
-                // 2. Plane where we have 1x Vector3 and 1x float field
-                // 3. Accesses of halves of larger SIMD types
-
-                if (indir->TypeIs(TYP_FLOAT))
+                bool simdElementAccessAllowed;
+                if (!m_compiler->IsBaselineSimdIsaSupported())
                 {
-                    if (((offset % genTypeSize(TYP_FLOAT)) == 0) && m_compiler->IsBaselineSimdIsaSupported())
-                    {
-                        return isDef ? IndirTransform::WithElement : IndirTransform::GetElement;
-                    }
+                    simdElementAccessAllowed = false;
                 }
-                else if (indir->TypeIs(TYP_SIMD12))
+                else if (indir->TypeIs(TYP_SIMD16))
                 {
-                    if ((offset == 0) && (varDsc->TypeGet() == TYP_SIMD16) && m_compiler->IsBaselineSimdIsaSupported())
-                    {
-                        return isDef ? IndirTransform::WithElement : IndirTransform::GetElement;
-                    }
+                    // This can be relaxed per-specific SIMD base type, but let's limit it to SSE4.1-x64 for simplicity.
+                    // It's required for WithElement + vector of longs.
+                    simdElementAccessAllowed = m_compiler->compOpportunisticallyDependsOn(InstructionSet_SSE41_X64);
+                }
+#if defined(TARGET_XARCH)
+                else if (indir->TypeIs(TYP_SIMD32))
+                {
+                    simdElementAccessAllowed = m_compiler->compOpportunisticallyDependsOn(InstructionSet_AVX);
+                }
+                else if (indir->TypeIs(TYP_SIMD64))
+                {
+                    simdElementAccessAllowed = m_compiler->compOpportunisticallyDependsOn(InstructionSet_AVX512F);
                 }
+#endif
```

Comment:
This doesn't look right. `indir->TypeIs(TYP_SIMD32)` means that we are going to use AVX512 intrinsics to extract/insert a half into an AVX512 register, so this case should check for AVX512. Similar for `indir->TypeIs(TYP_SIMD16)`.

Why does it need to be so complicated? Can `indir->TypeIs(TYP_SIMD16, TYP_SIMD32)` not just be changed to `(indir->TypeIs(TYP_SIMD16) && m_compiler->compOpportunisticallyDependsOn(InstructionSet_AVX)) || (indir->TypeIs(TYP_SIMD32) && m_compiler->compOpportunisticallyDependsOn(InstructionSet_AVX512F))`?

Reply:
@jakobbotsch ah, I misunderstood the logic here a bit, gotcha

Reply:
No worries -- I wonder if the logic could even just be `genTypeSize(varDsc) <= getMaxVectorByteLength()`

Reply:
Done!